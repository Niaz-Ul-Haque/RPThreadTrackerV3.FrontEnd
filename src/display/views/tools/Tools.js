import React, { Component } from 'react';
import {
	Row, Col, TabContent
} from 'reactstrap';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ToolsTabNav from './components/ToolsTabNav';
import ExportThreadsPane from './components/ExportThreadsPane';
import ManageTagsPane from './components/ManageTagsPane';
import { setActiveToolsTab, fetchTags } from '../../../infrastructure/actions';

const propTypes = {
	activeTab: PropTypes.string.isRequired,
	tags: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
	dispatch: PropTypes.func.isRequired
};

function mapStateToProps(state) {
	const {
		ui,
		user,
		tags
	} = state;
	return {
		user,
		tags,
		activeTab: ui.activeToolsTab
	};
}

class Tools extends Component {
	constructor(props) {
		super(props);
		this.setActiveTab = this.setActiveTab.bind(this);
	}
	componentDidMount() {
		const { dispatch } = this.props;
		if (!this.props.tags || !this.props.tags.length) {
			dispatch(fetchTags());
		}
	}
	setActiveTab(tab) {
		const { dispatch } = this.props;
		dispatch(setActiveToolsTab(tab));
	}
	render() {
		const { activeTab, tags } = this.props;
		return (
			<div className="animated fadeIn static-container settings-container">
				<Row>
					<Col>
						<ToolsTabNav setActiveTab={this.setActiveTab} activeTab={activeTab} />
						<TabContent activeTab={activeTab}>
							<ExportThreadsPane />
							<ManageTagsPane tags={tags} />
						</TabContent>
					</Col>
				</Row>
			</div>
		);
	}
}

Tools.propTypes = propTypes;
export default connect(mapStateToProps)(Tools);
