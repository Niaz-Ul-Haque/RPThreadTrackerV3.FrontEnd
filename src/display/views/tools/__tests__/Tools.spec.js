// #region imports
import React from 'react';
import { getSpecWrapper, shallowWithState } from '~/testhelpers/helpers.unit';
import Tools from '../Tools';
// #endregion imports

// #region mocks
jest.mock('../components/ManagePublicViewsPane', () => 'ManagePublicViewsPane');
jest.mock('../components/BrowserExtensionsPane', () => 'BrowserExtensionsPane');
jest.mock('../components/ExportThreadsPane', () => 'ExportThreadsPane');
jest.mock('../components/ManageTagsPane', () => 'ManageTagsPane');
jest.mock('../../../../infrastructure/actions', () => ({}));
jest.mock('../../../../infrastructure/selectors', () => ({
	getIsLoadingIconVisible: () => true,
	getIsManageTagsLoadingIconVisible: () => true
}));
jest.mock('../../../../infrastructure/constants/tabs', () => ({
	TOOLS: {
		TAB1: {
			href: '/tools/tab1',
			name: 'Tab 1'
		},
		TAB2: {
			tabId: '/tools/tab2',
			name: 'Tab 2'
		}
	}
}));
jest.mock('../../../shared/static/StaticTabNav', () => 'StaticTabNav');
jest.mock('../../../shared/static/StaticDropdownNav', () => 'StaticDropdownNav');
// #endregion mocks

const createTestProps = propOverrides => ({
	fetchCharacters: jest.fn(),
	fetchTags: jest.fn(),
	fetchUser: jest.fn(),
	fetchPublicViews: jest.fn(),
	setActiveToolsTab: jest.fn(),
	exportThreads: jest.fn(),
	openUpsertPublicViewModal: jest.fn(),
	openDeletePublicViewModal: jest.fn(),
	openBulkUpdateTagModal: jest.fn(),
	openBulkDeleteTagModal: jest.fn(),
	match: {
		url: '/tools/tab1',
		params: { tabId: 'tab1' }
	},
	...propOverrides
});

const createTestState = stateOverrides => ({
	user: { id: '12345', userName: 'test-user' },
	tags: ['tag1', 'tag2', 'tag3'],
	publicViews: [{}, {}, {}, {}],
	characters: [{}, {}],
	ui: { isBulkUpdateTagModalOpen: true, isBulkDeleteTagModalOpen: true },
	...stateOverrides
});

describe('rendering', () => {
	describe('snapshots', () => {
		it('should render valid snapshot', () => {
			const props = createTestProps();
			const state = createTestState();
			const jsx = (<Tools {...props} />);
			const element = shallowWithState(jsx, state).dive();
			expect(element).toMatchSnapshot();
		});
	});
	describe('options', () => {
		it('should populate options from imported tabs on dropdown nav', () => {
			const props = createTestProps();
			const state = createTestState();
			const jsx = (<Tools {...props} />);
			const element = shallowWithState(jsx, state).dive();
			const form = getSpecWrapper(element, 'tools-static-dropdown-nav');
			const { options } = form.props();
			expect(options).toHaveLength(2);
		});
		it('should populate options from imported tabs on tab nav', () => {
			const props = createTestProps();
			const state = createTestState();
			const jsx = (<Tools {...props} />);
			const element = shallowWithState(jsx, state).dive();
			const form = getSpecWrapper(element, 'tools-static-tab-nav');
			const { options } = form.props();
			expect(options).toHaveLength(2);
		});
	});
});

describe('behavior', () => {
	describe('componentDidMount', () => {
		it('should retrieve tags when tags are not loaded', () => {
			const fetchTags = jest.fn();
			const props = createTestProps({ fetchTags });
			const state = createTestState({ tags: [] });
			const jsx = (<Tools {...props} />);
			shallowWithState(jsx, state).dive();
			expect(fetchTags).toHaveBeenCalledTimes(1);
		});
		it('should not retrieve tags when tags are loaded', () => {
			const fetchTags = jest.fn();
			const props = createTestProps({ fetchTags });
			const state = createTestState();
			const jsx = (<Tools {...props} />);
			shallowWithState(jsx, state).dive();
			expect(fetchTags).toHaveBeenCalledTimes(0);
		});
		it('should retrieve public views when public views are not loaded', () => {
			const fetchPublicViews = jest.fn();
			const props = createTestProps({ fetchPublicViews });
			const state = createTestState({ publicViews: [] });
			const jsx = (<Tools {...props} />);
			shallowWithState(jsx, state).dive();
			expect(fetchPublicViews).toHaveBeenCalledTimes(1);
		});
		it('should not retrieve public views when public views are loaded', () => {
			const fetchPublicViews = jest.fn();
			const props = createTestProps({ fetchPublicViews });
			const state = createTestState();
			const jsx = (<Tools {...props} />);
			shallowWithState(jsx, state).dive();
			expect(fetchPublicViews).toHaveBeenCalledTimes(0);
		});
		it('should retrieve characters when characters are not loaded', () => {
			const fetchCharacters = jest.fn();
			const props = createTestProps({ fetchCharacters });
			const state = createTestState({ characters: [] });
			const jsx = (<Tools {...props} />);
			shallowWithState(jsx, state).dive();
			expect(fetchCharacters).toHaveBeenCalledTimes(1);
		});
		it('should not retrieve characters when characters are loaded', () => {
			const fetchCharacters = jest.fn();
			const props = createTestProps({ fetchCharacters });
			const state = createTestState();
			const jsx = (<Tools {...props} />);
			shallowWithState(jsx, state).dive();
			expect(fetchCharacters).toHaveBeenCalledTimes(0);
		});
		it('should retrieve user when user is not loaded', () => {
			const fetchUser = jest.fn();
			const props = createTestProps({ fetchUser });
			const state = createTestState({ user: {} });
			const jsx = (<Tools {...props} />);
			shallowWithState(jsx, state).dive();
			expect(fetchUser).toHaveBeenCalledTimes(1);
		});
		it('should not retrieve user when user is loaded', () => {
			const fetchUser = jest.fn();
			const props = createTestProps({ fetchUser });
			const state = createTestState({ user: { id: '12345', userName: 'my-user' } });
			const jsx = (<Tools {...props} />);
			shallowWithState(jsx, state).dive();
			expect(fetchUser).toHaveBeenCalledTimes(0);
		});
	});
	describe('onExportRequest', () => {
		it('should trigger export action', () => {
			const exportThreads = jest.fn();
			const props = createTestProps({ exportThreads });
			const state = createTestState();
			const jsx = (<Tools {...props} />);
			const element = shallowWithState(jsx, state).dive();
			element.instance().onExportRequest(true, true);
			expect(exportThreads).toHaveBeenCalledTimes(1);
			expect(exportThreads).toHaveBeenLastCalledWith({
				includeHiatused: true, includeArchive: true
			});
		});
	});
});
