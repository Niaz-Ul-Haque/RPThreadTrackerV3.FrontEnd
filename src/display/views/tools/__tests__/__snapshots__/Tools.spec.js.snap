// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering snapshots should render valid snapshot 1`] = `
<div
  className="animated fadeIn static-container settings-container"
>
  <Row
    tag="div"
  >
    <Col
      className="d-lg-none text-center"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <StaticDropdownNav
        activeTab="my-test-tab"
        data-spec="tools-static-dropdown-nav"
        options={
          Array [
            Object {
              "name": "Tab 1",
              "tabId": "tab1",
            },
            Object {
              "name": "Tab 2",
              "tabId": "tab2",
            },
          ]
        }
        setActiveTab={[MockFunction]}
      />
    </Col>
  </Row>
  <Row
    tag="div"
  >
    <Col
      className="d-none d-lg-block"
      md={3}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <StaticTabNav
        activeTab="my-test-tab"
        data-spec="tools-static-tab-nav"
        options={
          Array [
            Object {
              "name": "Tab 1",
              "tabId": "tab1",
            },
            Object {
              "name": "Tab 2",
              "tabId": "tab2",
            },
          ]
        }
        setActiveTab={[MockFunction]}
      />
    </Col>
    <Col
      lg="9"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
      xs="12"
    >
      <TabContent
        activeTab="my-test-tab"
        tag="div"
      >
        <ExportThreadsPane
          onExportRequest={[Function]}
        />
        <ManagePublicViewsPane
          isLoadingIconVisible={true}
          openDeletePublicViewModal={[MockFunction]}
          openUpsertPublicViewModal={[MockFunction]}
          publicViews={
            Array [
              Object {},
              Object {},
              Object {},
              Object {},
            ]
          }
          username="test-user"
        />
        <BrowserExtensionsPane />
      </TabContent>
    </Col>
  </Row>
</div>
`;
